core/src/kvs/surrealkv/mod.rs

```diff
@@ -0,0 +1,398 @@
+
+use std::ops::Range;
+use surrealkv::Options;
+use surrealkv::Store;
+use surrealkv::Transaction as Tx;
+
+pub struct Datastore {
+	db: Store,
+}
+
+pub struct Transaction {
+	/// Is the transaction complete?
+	done: bool,
+	/// Is the transaction writeable?
+	write: bool,
+	/// Should we check unhandled transactions?
+	check: Check,
+	/// The underlying datastore transaction
+	inner: Tx,
+}
+
+impl Drop for Transaction {
+	fn drop(&mut self) {
+		if !self.done && self.write {
+			// Check if already panicking
+			if std::thread::panicking() {
+				return;
+			}
+			// Handle the behaviour
+			match self.check {
+				Check::None => {
+					trace!("A transaction was dropped without being committed or cancelled");
+				}
+				Check::Warn => {
+					warn!("A transaction was dropped without being committed or cancelled");
+				}
+				Check::Panic => {
+					#[cfg(debug_assertions)]
+					{
+						let backtrace = std::backtrace::Backtrace::force_capture();
+						if let std::backtrace::BacktraceStatus::Captured = backtrace.status() {
+							println!("{}", backtrace);
+						}
+					}
+					panic!("A transaction was dropped without being committed or cancelled");
+				}
+			}
+		}
+	}
+}
+
+impl Datastore {
+	/// Open a new database
+	pub(crate) async fn new(path: &str) -> Result<Datastore, Error> {
+		let mut opts = Options::new();
+		opts.dir = path.to_string().into();
+
+		match Store::new(opts) {
+			Ok(db) => Ok(Datastore {
+				db,
+			}),
+			Err(e) => Err(Error::Ds(e.to_string())),
+		}
+	}
+	/// Start a new transaction
+	pub(crate) async fn transaction(&self, write: bool, _: bool) -> Result<Transaction, Error> {
+		// Specify the check level
+		#[cfg(not(debug_assertions))]
+		let check = Check::Warn;
+		#[cfg(debug_assertions)]
+		let check = Check::Panic;
+		// Create a new transaction
+		match self.db.begin() {
+			Ok(inner) => Ok(Transaction {
+				done: false,
+				check,
+				write,
+				inner,
+			}),
+			Err(e) => Err(Error::Tx(e.to_string())),
+		}
+	}
+}
+
+impl Transaction {
+	/// Behaviour if unclosed
+	pub(crate) fn check_level(&mut self, check: Check) {
+		self.check = check;
+	}
+	/// Check if closed
+	pub(crate) fn closed(&self) -> bool {
+		self.done
+	}
+	/// Cancel a transaction
+	pub(crate) async fn cancel(&mut self) -> Result<(), Error> {
+		// Check to see if transaction is closed
+		if self.done {
+			return Err(Error::TxFinished);
+		}
+		// Mark this transaction as done
+		self.done = true;
+
+		// Cancel this transaction
+		self.inner.rollback();
+
+		// Continue
+		Ok(())
+	}
+	/// Commit a transaction
+	pub(crate) async fn commit(&mut self) -> Result<(), Error> {
+		// Check to see if transaction is closed
+		if self.done {
+			return Err(Error::TxFinished);
+		}
+		// Check to see if transaction is writable
+		if !self.write {
+			return Err(Error::TxReadonly);
+		}
+		// Mark this transaction as done
+		self.done = true;
+
+		// Commit this transaction
+		self.inner.commit().await?;
```

Comment:
Haven't checked API, but does commit return `Result<(), Error>`? If so, can just return that instead

Reply:
Thank you for suggesting this

Reply:
Just checked, it doesn't directly return it